autoprotocol/builders.py

```diff
@@ -1738,6 +1745,154 @@ def blowout(self, volume, initial_z, flowrate=None):
+                        reference="preceding_position"
+                    )
+                )
+            ),
+            LiquidHandle.builders.transport(
+                volume=Unit(1, "uL"),
+                density=None,
+                pump_override_volume=Unit(2, "uL"),
+                flowrate=LiquidHandle.builders.flowrate(
+                    target=Unit(10, "uL/s")
+                ),
+                delay_time=Unit(0.5, "s"),
+                mode_params=LiquidHandle.builders.mode_params(
+                    liquid_class="viscous",
+                    position_z=LiquidHandle.builders.position_z(
+                        reference="preceding_position"
+                    )
+                )
+            )
+        ]
+
+        LiquidHandle.builders.desired_mode(example_transports, None)
+
+        This will return "positive_displacement" based on the "viscous" liquid
+        class. Note that "air" (which defaults to "air_displacement") does not
+        cause a conflict since "air" is often applied in additional transfers
+        of air (such as blowout) for any liquid class, and is disregarded.
+
+        Parameters
+        ----------
+        transports : dict, optional
+            Dictionary of the transport parameters
+        mode : str, optional
+            Mode of dispense type
+
+        Returns
+        -------
+        str
+            Mode of dispense type
+
+        Raises
+        ------
+        TypeError
+            mode is not in str
+        ValueError
+            mode does not contain valid mode value
+        TypeError
+            liquid is not of str
+        ValueError
+            liquid_class is not a valid value
+        ValueError
+            liquid_class did not resolve to a mode
+        ValueError
+            multiple liquid_class exists in one LiquidHandle
+        """
+        # dict for default dispense mode for each liquid_class
+        liquid_class_to_dispense_modes = {
+            "air": "air_displacement",
+            "default": "air_displacement",
+            "viscous": "positive_displacement"
+        }
+        if mode:
+            if not isinstance(mode, str):
+                raise TypeError(
+                    "mode: {} must be in str".format(mode)
+                )
+            if mode not in self.dispense_modes:
+                raise ValueError(
+                    "mode: {} must be one of the valid modes: {}"
+                    "".format(mode, self.dispense_modes)
+                )
+        else:
+            # get liquid_classes from transport input
+            liquid_classes = set([transport["mode_params"]["liquid_class"]
+                                  for transport in transports])
+            # remove automatically added 'air' class from blowout, etc.
+            non_air_classes = [liq for liq in liquid_classes if liq != "air"]
+            # if liquid classes only contain "air" type, mode is returned.
+            if not non_air_classes:
```

Comment:
Double-negative here ("not non-...") followed by a mid-method return. Can we try to improve this without adding additional nesting? Creating a new method is fine.

Reply:
Hmm. Not the best way to state 'if there is no non-air classes' here... I think it is easier than checking for 'transports that only contain 'air' and nothing else' (and state 'if it only contains air'), since I'll have to check for what other liquid_class is present in a list of transports later... I can change variable name if 'non-air...' doesn't read right/weird? Also, I think it's easier to just return the value here since if transports only contain `air` no further logic is necessary. I can move this to a later point, but I'll probably need to squeeze in another nested `if` anyways, since if transports only contain `air`,`modes` should be treated differently...

Reply:
I mean, is this statement even needed?

If `non_air_classes` is empty, the for loop will skip and we will go directly to 1886, which has the same effect as returning immediately with `liquid_class_to_dispense_modes["air"]`.

```
non_air_classes = [liq for liq in liquid_classes if liq != "air"]
// non_air_classes = []
 modes = []
 for liquid in non_air_classes: //skipped
    ...
if not modes: // True, line 1186
   modes.append(liquid_class_to_dispense_modes["air"])
```

Reply:
This `modes` will also be empty if everything else in the `transports` is `None`. I like to have different behavior for when it contains `air` vs `None`, so I'll still need to have another logic here to separate the two. I don't mind changing if it's easier to read, tho. I don't have a strong preference one way or another... I'll think about it a bit more to see if i can make this better. Meanwhile, let me know if you have suggestions or strong preference for how it reads!

Reply:
Hmm, I'm not following. Can we meet to discuss?

Reply:
You say different behavior, but it's the same behavior, right? The same effect...

Reply:
Removed double-negative condition :)